backwards melisma

MarcSabatella opened the pull request:
There are a few situations where we draw melisma lines backwards or otherwise incorrectly, this PR should fix them.

- if the syllable with the melisma extends past the right edge of the next note (https://musescore.org/en/node/257131)

- if another voice has an earlier note that ends before the end note of this voice (https://musescore.org/en/node/282099)

The first fix is a very simple change that should prevent any and all backwards melisma lines, leaving the zero length instead.  The second change is bit more involved, it fundamentally changes the calculation of the end element for melisma lines.  As such, it definitely warrants further review/testing.  In the basic cases it should always result in the same thing, but it's the overlapping voices where it changes, hopefully always for the better.

Jojo-Schmitz commented on libmscore/lyricsline.cpp:
```diff
@@ -183,7 +183,7 @@ SpannerSegment* LyricsLine::layoutSystem(System* system)
             layout();
             if (ticks().isZero())                 // only do layout if some time span
-                  return 0;
+                  return nullptr;
```
Should this be made part of the [MuseScore coding style](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)?
It isn't strictly needed, as a 0 in any pointer context automagically turns into a nullptr (C++) or NULL (C), but might be better doe selfdocumenting code?

MarcSabatella replied:
Maybe so.  I know it gets mentioned in code reviews, so I just think of it as part of the style even if it is not officially.

Jojo-Schmitz replied:
OK, now added in https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules#Returning-pointers

dmitrio95 commented on libmscore/spanner.cpp:
```diff
@@ -596,13 +597,35 @@ void Spanner::computeEndElement()
+                        // lyrics ticks should already indicate the segment we want
+                        // except for TEMP_MELISMA_TICKS case
+                        Lyrics* l = toLyricsLine(this)->lyrics();
+                        Fraction tick = (l->ticks().ticks() == Lyrics::TEMP_MELISMA_TICKS) ? l->tick() : l->endTick();
```
To make it more clear, was the issue that `LyricsLine::tick2()` returned a wrong tick value or was it `findCRinStaff` that returned a chord from a wrong segment?

MarcSabatella replied:
Well, findCRinStaff() does what it supposed to do, it's the right thing for most line types, it's just the wrong thing for melisma lines which have their own unique rules.  There is no value we can pass that function that will produce what we want, so no way to improve the result just by altering LyricsLine::tick2().  Although now that you mention it, it probably would be better to use that value here instead of Lyrics::tick2().  They aren't necessarily the same, but probably they should be.  I will take another look at this.

MarcSabatella commented:
I updated the PR just to add some comments & fix a debug message.  I tried making the lyricsline tick2 same as the lyrics, but it just isn't designed to work that way - the lyricsline tick2 is for the *next* note, same as for all other lines.  So I do need to continue to use lyrics tick2.

Meanwhile, I added a vtest to show the two cases that are wrong.  Attempts to use findCRinStaff() with an appropriate tick value can fix one but not the other, depending on hat tick you pass in.

![image](https://user-images.githubusercontent.com/1799936/56516826-b605a100-64f8-11e9-9774-0d6d9fd3246b.png)

dmitrio95 commented:
OK, thanks for the explanation! Let's use this solution then.
